src/levanter/infra/ray_tpu.py

```diff
@@ -63,73 +63,85 @@ class TpuFailed(_TpuRunResult):
+            self.num_hosts = ray.util.accelerators.tpu.get_current_pod_worker_count()
+            self.ip = socket.gethostbyname(socket.gethostname())
+            self.slice_actors = None
 
-    Args:
-        remote_fn: A remote function that takes no arguments
-        tpu_type: The type of TPU to run on, e.g. "v4-32"
-        num_slices: The number of slices to run
+        def get_slice_info(self):
+            return self.pod_name, self.num_hosts, self.ip
+        
+        def run(self, remote_fn, num_slices) -> _TpuRunResult:
+            if self.slice_actors is not None:
+                raise RuntimeError("Actors already created")
+            
+            SliceActor = make_tpu_slice_actor(tpu_type)
+            self.slice_actors = [SliceActor.remote() for _ in range(num_slices)]
+
+            # Get info from all slice actors to ensure they're ready
+            futures = [actor.get_slice_info.remote() for actor in self.slice_actors]
+            try:
+                slice_infos = ray.get(futures)
+                logger.info(f"TPU slice infos: {slice_infos}")
+            except Exception as e:
+                self.kill()
+                raise RuntimeError("Failed to initialize slice actors") from e
+            
+            # Start process on all slices
+            try:
+                futures = [
+                    actor.run.remote(remote_fn, self.ip, i, num_slices) 
+                    for i, actor in enumerate(self.slice_actors)
+                ]
+                infos = ray.get(futures)
 
-    Returns:
-        A Ray ObjectRef that represents the result of the function
-    """
+                wait_futures = [actor.wait.remote() for actor in self.slice_actors]
+                results = ray.get(wait_futures)
 
-    @ray.remote(resources={f"TPU-{tpu_type}-head": 1})
-    class MultisliceActor:
+                logger.info("TPU job finished, with info", infos[0])
+                return TpuSuccess(infos[0], results)
+            except RayError as e:
+                logger.exception(f"Failed to run job on {self.pod_name}")
+                self.kill()
+                return _handle_ray_error(infos[0], e)
+            except Exception as e:
+                logger.exception(f"Failed to run job on {self.pod_name}")
+                self.kill()
+                raise RuntimeError(f"Failed to run job on {self.pod_name}") from e
+            
+        def kill(self):
+            if self.slice_actors is not None:
+                for actor in self.slice_actors:
+                    try:
+                        ray.kill(actor)
+                    except Exception:
+                        logger.exception(f"Failed to kill slice actor {actor}")
+                self.slice_actors = None
+            ray.actor.exit_actor()
+    
+    return TPUHeadNodeActor
+
+def make_tpu_slice_actor(tpu_type: str):
+    @ray.remote(resources={f"TPU-{tpu_type}-slice": 1})
```

Comment:
this isn't a thing. the relevant resource types are:

* `TPU-{tpu_type}-head`: special resource given to the 0th worker of a tpu pod slice
* `{name of tpu slice}`: special resource given to all workers in a particular tpu slice
* (plus the TPU resources for a given tpu worker)

So the types of actors are:

* 1 head node actor for each tpu slice
* K worker actors owned by the head node actor

(A multislice job will have multiple HeadActors)